scripts/downgrade.sh

```diff
@@ -0,0 +1,197 @@
+#!/bin/bash
+set -e
+set -u
+set -o pipefail
+
+VERBOSE=${VERBOSE:-0}
+LOGFILE=/var/log/metalk8s-downgrade.log
+DRY_RUN=0
+SALTENV=""
+DESTINATION_VERSION=""
+SALT=""
+SALT_CALL=${SALT_CALL:-salt-call}
+CRICTL=${CRICTL:-crictl}
+
+
+_usage() {
+    echo "downgrade.sh [options]"
+    echo "Options:"
+    echo "--destination-version "
+    echo "   <destination-version>:        Destination version to downgrade to"
+    echo "-l/--log-file <logfile_path>:    Path to log file"
+    echo "-v/--verbose:                    Run in verbose mode"
+    echo "-d/--dry-run:                    Run actions in dry run mode"
+    echo "-h/--help:                       Show this help menu"
+}
+
+
+while (( "$#" )); do
+  case "$1" in
+    -d|--dry-run)
+      DRY_RUN=1
+      shift
+      ;;
+    -v|--verbose)
+      VERBOSE=1
+      shift
+      ;;
+    --destination-version)
+      DESTINATION_VERSION="$2"
+      shift 2
+      ;;
+    -l|--log-file)
+      LOGFILE="$2"
+      shift 2
+      ;;
+    -h|--help)
+      _usage
+      exit
+      ;;
+    *) # unsupported flags
+      echo "Error: Unsupported flag $1" >&2
+      _usage
+      exit 1
+      ;;
+  esac
+done
+
+TMPFILES=$(mktemp -d)
+
+cat << EOF >> "${LOGFILE}"
+--- MetalK8s downgrade started on $(date -u -R) ---
+EOF
+
+exec > >(tee -ia "${LOGFILE}") 2>&1
+
+cleanup() {
+    rm -rf "${TMPFILES}" || true
+}
+
+trap cleanup EXIT
+
+run_quiet() {
+    local name=$1
+    shift 1
+
+    echo -n "> ${name}..."
+    local start
+    start=$(date +%s)
+    set +e
+    "$@" 2>&1 | tee -ia "${LOGFILE}" > "${TMPFILES}/out"
+    local RC=$?
+    set -e
+    local end
+    end=$(date +%s)
+
+    local duration=$(( end - start ))
+
+    if [ $RC -eq 0 ]; then
+        echo " done [${duration}s]"
+    else
+        echo " fail [${duration}s]"
+        cat >/dev/stderr << EOM
+
+Failure while running step '${name}'
+
+Command: $@
+
+Output:
+
+<< BEGIN >>
+EOM
+        cat "${TMPFILES}/out" > /dev/stderr
+
+        cat >/dev/stderr << EOM
+<< END >>
+
+This script will now exit
+
+EOM
+
+        exit 1
+    fi
+}
+
+run_verbose() {
+    local name=$1
+    shift 1
+
+    echo "> ${name}..."
+    "$@"
+}
+
+run() {
+    if [ "$VERBOSE" -eq 1 ]; then
+        run_verbose "${@}"
+    else
+        run_quiet "${@}"
+    fi
+}
+
+die() {
+    echo 1>&2 "$@"
+    return 1
+}
+
+_set_env() {
+    if [ -z "$SALTENV" ]; then
+        SALTENV="metalk8s-$($SALT_CALL --out txt slsutil.renderer \
```

Comment:
Hmm here you get the current bootstrap version but what we want it's the highest version installed and available but it's not really simple to get this one especially in a bash script :/ We should have an external pillar that will retrieve this. But maybe lets take the bootstrap version for this PR

Reply:
Yes this one should be rethought, lets create a ticket

Reply:
Btw we should also not downgrade/upgrade the bootstrap if it's already older/newer than the destination version. But not in these two PR I think

Reply:
that should be in the precheck no ?

Reply:
That can't because we can downgrade/upgrade just we should skip the node already older/newer